flask/app/variants.py

```diff
@@ -0,0 +1,424 @@
+            .join(models.Dataset.genotype)
+            .join(models.Genotype.analysis, models.Genotype.variant)
+            .join(models.Variant.gene)
+            .join(models.Participant.family)
+            .join(
+                models.groups_datasets_table,
+                models.Dataset.dataset_id
+                == models.groups_datasets_table.columns.dataset_id,
+            )
+            .join(
+                models.users_groups_table,
+                models.groups_datasets_table.columns.group_id
+                == models.users_groups_table.columns.group_id,
+            )
+            .options(
+                contains_eager(models.Dataset.genotype)
+                .contains_eager(models.Genotype.variant)
+                .contains_eager(models.Variant.gene),
+                contains_eager(models.Dataset.tissue_sample)
+                .contains_eager(models.TissueSample.participant)
+                .contains_eager(models.Participant.family),
+            )
+            .filter(or_(*filters), models.users_groups_table.columns.user_id == user_id)
+        )
+
+    else:
+        app.logger.debug("Processing query - unrestricted based on user id.")
+
+        query = (
+            db.session.query(models.Dataset)
+            .join(models.Participant.tissue_samples)
+            .join(models.TissueSample.datasets)
+            .join(models.Dataset.genotype)
+            .join(models.Genotype.analysis, models.Genotype.variant)
+            .join(models.Variant.gene)
+            .join(models.Participant.family)
+            .options(
+                contains_eager(models.Dataset.genotype)
+                .contains_eager(models.Genotype.variant)
+                .contains_eager(models.Variant.gene),
+                contains_eager(models.Dataset.tissue_sample)
+                .contains_eager(models.TissueSample.participant)
+                .contains_eager(models.Participant.family),
+            )
+            .filter(or_(*filters))
+        )
+
+    app.logger.info(request.accept_mimetypes)
+
+    if request.accept_mimetypes["application/json"]:
+
+        app.logger.info("Defaulting to json response")
+
+        q = query.all()
+        app.logger.info(
+            [
+                {
+                    "participant_codename": dataset.tissue_sample.participant.participant_codename,
+                    "family_codename": dataset.tissue_sample.participant.family.family_codename,
+                }
+                for dataset in q
+            ]
+        )
+        return jsonify(
+            [
+                {
+                    "participant_codename": dataset.tissue_sample.participant.participant_codename,
+                    "family_codename": dataset.tissue_sample.participant.family.family_codename,
+                    "dataset": [
+                        {
+                            "variants": [
+                                {
+                                    **asdict(genotype),
```

Comment:
I see the semantic utility of combining properties from the gene, variant, and genotype models here, but I also wonder what it means for programmatic consumers of the response. For instance, the front-end might need a transformer and intermediate Typescript definitions to 'rebuild' the underlying models so that they  work with existing code. But maybe these concerns are irrelevant for `summary/` endpoints, which by definition give a 'flattened' view of the data.

Reply:
I came up with this idea but it's open to change later as we don't really know what a frontend consumer might do with these yet.